fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,482 @@
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#include "ast.h"
+#include "ast.hpp"
+#include <stdio.h>
+#include <string>
+#include <iostream>
+#include <vector>
+#include <iterator>
+
+
+static ASTNode* trees;
+
+int MAX_NUM = 128;
+
+class Token {
+  public:
+    enum Type {
+        ID = 0,
+        STRING_,
+        STRING_0,
+        INT,
+        WHITESPACE,
+        AT,
+        ZERO,
+        HEXINT,
+        QUOT,
+        LPAREN,
+        RPAREN,
+        LPAREN_SQ,
+        RPAREN_SQ,
+        BECOMES,
+        COMMA,
+        COMMENT,
+        DOUBLE_
+    };
+
+    private:
+        Type type;
+        std::string value;
+    
+
+    public:
+        Token(Type type, std::string value):
+	        type(type), value(value) {}
+        Type getType() const { return type; }
+        void setType(Token::Type t) {type = t ;}
+        const std::string &getValue() const {return value ;}
+};
+
+class TokenFailure {
+	std::string message;
+
+public:
+	TokenFailure(std::string message);
+	const std::string &what() const;
+};
+
+TokenFailure::TokenFailure(std::string message) :
+	message(message) {}
+
+const std::string &TokenFailure::what() const { return message; }
+
+int isNotNewLine(int c) { return c != '\n'; }
+
+int isNotQuot(int c) { return c != '"'; }
+
+int isHex(int c) {
+    if (c == 'a' || c == 'b' || c == 'c' || c == 'd' || c == 'e' || c =='f' ||
+        c == 'A' || c == 'B' || c == 'C' || c == 'D' || c == 'E' || c == 'F') {
+        return true;
+    } else {
+        return false;
+    }
+}
+
+int isNotZero(int c) { return c != '0' && isdigit(c); }
+
+class ASTstates {
+    public:
+        enum State {
+            ID = 0,
+            STRING_,
+            STRING_0,
+            INT,
+            WHITESPACE,
+            AT,
+            ZERO,
+            HEXINT,
+            QUOT,
+            PAREN,
+            LPAREN,
+            RPAREN,
+            LPAREN_SQ,
+            RPAREN_SQ,
+            BECOMES,
+            COMMA,
+            DOUBLE_,
+
+            START,
+            FAIL,
+            DOT,
+            MINUS,
+            COMMENT,
+            LARGEST_STATE = COMMENT
+        };
+    
+    private:
+        std::vector<State> acceptingStates;
+        std::vector<std::vector<State> > transitionFunction;
+        Token::Type stateToType(State s) const {
```

Comment:
Could a `const std::unordered_map<State, Token::Type>` do the trick here?

Reply:
Yes, it could. Unfortunately, I don't believe we have `std::unordered_map`on all our build platforms.